iomem_pci.c

```diff
@@ -0,0 +1,326 @@
+{
+	int err;
+
+	err = iomem_test_page_mappings(p);
+	if (err)
+		return err;
+
+	return iomem_test_p2p_access(p);
+}
+
+static void iomem_release(struct device *dev)
+{
+	struct iomem_dev *p = to_iomem(dev);
+
+	kfree(p);
+}
+
+static struct iomem_dev *iomem_create(struct pci_dev *pdev)
+{
+	struct iomem_dev *p;
+	int err;
+
+	p = kzalloc(sizeof(*p), GFP_KERNEL);
+	if (!p)
+		return ERR_PTR(-ENOMEM);
+
+	p->pdev = pdev;
+
+	device_initialize(&p->dev);
+	p->dev.class = iomem_class;
+	p->dev.parent = &pdev->dev;
+	p->dev.release = iomem_release;
+
+	p->id = ida_simple_get(&iomem_ida, 0, 0, GFP_KERNEL);
+	if (p->id < 0) {
+		err = p->id;
+		goto out_free;
+	}
+
+	dev_set_name(&p->dev, "iomem%d", p->id);
+	p->dev.devt = MKDEV(MAJOR(iomem_devt), p->id);
+
+	cdev_init(&p->cdev, &iomem_fops);
+	p->cdev.owner = THIS_MODULE;
+
+	err = cdev_device_add(&p->cdev, &p->dev);
+	if (err)
+		goto out_ida;
+
+	dev_info(&p->dev, "registered");
+
+	iomem_test(p);
+
+	return p;
+
+out_ida:
+	ida_simple_remove(&iomem_ida, p->id);
+out_free:
+	kfree(p);
+	return ERR_PTR(err);
+}
+
+void iomem_destroy(struct iomem_dev *p)
+{
+	dev_info(&p->dev, "unregistered");
+	cdev_device_del(&p->cdev, &p->dev);
+	ida_simple_remove(&iomem_ida, p->id);
+	put_device(&p->dev);
+}
+
+static int iomem_pci_probe(struct pci_dev *pdev,
+			    const struct pci_device_id *id)
+{
+	struct iomem_dev *p;
+	int err = 0;
+
+	if (pci_enable_device_mem(pdev) < 0) {
+		dev_err(&pdev->dev, "unable to enable device!\n");
+		goto out;
+	}
+
+	err = pci_p2pdma_add_resource(pdev, 0, 0, 0);
+	if (err) {
+		dev_err(&pdev->dev, "unable to add p2p resource");
+		goto out_disable_device;
+	}
+
+	pci_p2pmem_publish(pdev, true);
+
+	p = iomem_create(pdev);
+	if (IS_ERR(p))
+		goto out_disable_device;
+
+	pci_set_drvdata(pdev, p);
+
+	return 0;
+
+out_disable_device:
+	pci_disable_device(pdev);
+out:
+	return err;
+}
+
+static void iomem_pci_remove(struct pci_dev *pdev)
+{
+	struct iomem_dev *p = pci_get_drvdata(pdev);
+
+	iomem_destroy(p);
+}
+
+static struct pci_driver iomem_pci_driver = {
+	.name = "iomem_pci",
+	.id_table = iomem_pci_id_table,
+	.probe = iomem_pci_probe,
+	.remove = iomem_pci_remove,
+};
+
+static int __init iomem_pci_init(void)
+{
+	int rc;
+
+	iomem_class = class_create(THIS_MODULE, "iomem_device");
+	if (IS_ERR(iomem_class))
+		return PTR_ERR(iomem_class);
+
+	rc = alloc_chrdev_region(&iomem_devt, 0, max_devices, "iomem");
+	if (rc)
+		goto err_class;
+
+	rc = pci_register_driver(&iomem_pci_driver);
+	if (rc)
+		goto err_chdev;
+
+	pr_info(KBUILD_MODNAME ": module loaded\n");
+
+	return 0;
+err_chdev:
+	unregister_chrdev_region(iomem_devt, max_devices);
+err_class:
+	class_destroy(iomem_class);
+	return rc;
+}
+
+static void __exit iomem_pci_cleanup(void)
+{
+	pci_unregister_driver(&iomem_pci_driver);
+	unregister_chrdev_region(iomem_devt, max_devices);
+	class_destroy(iomem_class);
+	pr_info(KBUILD_MODNAME ": module unloaded\n");
+}
+
+late_initcall(iomem_pci_init);
```

Comment:
I suspect we don't need this anymore. I think that was only necessary for the horid hack of the driver you copied.

Reply:
Yup it seems to work normally now.